File: minion-agent-python/src/minion_agent/agent/instance.py
"""One live execution identity: inbox, log, turn state, scope."""

from __future__ import annotations

from collections.abc import Callable

from ..llm import Message, ModelId
from ..runtime import Context, ScopeKey
from ..session import SessionLog, derive_messages
from .envelope import InboxTarget, InputEnvelope, JsonValue
from .events import AGENT_STATUS, declare_agent_events
from .identity import AgentDefinition, AgentInstanceId, AgentStatus, ThinkingLevel
from .inbox import Inbox


class AgentActiveError(RuntimeError):
    """An operation that requires an idle instance was attempted while running."""


def instance_scope_key(definition: AgentDefinition, instance_id: str) -> ScopeKey:
    """The scope an instance registers through.

    A child of the definition's scope, so definition-level registrations are
    visible to every instance while instance-level ones stay private to one
    (design spec section 3).
    """
    return ScopeKey(f"agent-instance:{instance_id}", parent=ScopeKey(definition.scope_name))


class AgentInstance:
    """One conversation with one agent."""

    def __init__(
        self,
        *,
        instance_id: AgentInstanceId,
        definition: AgentDefinition,
        log: SessionLog,
        ctx: Context,
    ) -> None:
        self.id = instance_id
        self.definition = definition
        self.log = log
        self.inbox = Inbox()
        self._status = AgentStatus.IDLE
        self.on_status_change: Callable[[AgentStatus], None] | None = None

        # Mutable per-instance current configuration (`AG-014`, `L07-R001`):
        # pinned Pi's `AgentState.systemPrompt`/`model`/`thinkingLevel` are directly
        # assignable and read for later runs -- distinct from `definition`'s own
        # shared, frozen defaults, which never change once an instance has its own
        # current value. Defaults to the definition's own value, matching pinned
        # Pi's `createMutableAgentState`'s `initialState?.systemPrompt ?? ""` /
        # `initialState?.model ?? DEFAULT_MODEL` pattern.
        self.system_prompt: str = definition.system
        self.model: ModelId = definition.model
        self.thinking_level: ThinkingLevel = ThinkingLevel.OFF

        # Runtime-state vocabulary (`AG-015`): pinned Pi's `streamingMessage`/
        # `pendingToolCalls`/`errorMessage`, initial values only -- Layer 08 owns
        # when and how these change during a run; this layer owns only that they
        # exist and what they start as.
        self.streaming_message: Message | None = None
        self.pending_tool_calls: frozenset[str] = frozenset()
        self.error_message: str | None = None

        declare_agent_events(ctx.events)
        self.scope = ctx.scope(instance_scope_key(definition, instance_id))
        self._ctx = self.scope.ctx

    @property
    def ctx(self) -> Context:
        """This instance's scoped context."""
        return self._ctx

    @property
    def status(self) -> AgentStatus:
        return self._status

    @property
    def messages(self) -> tuple[Message, ...]:
        """The current transcript (`AG-015`): a fresh projection of the
        authoritative `SessionLog` (Layer 03, certified), never a live mutable
        reference -- an intentional divergence from pinned Pi's own
        `state.messages` getter (which returns the live backing array) since
        Session's log is the sole authority for history and must not be
        corruptible by mutating a returned collection in place."""
        return derive_messages(self.log)

    def set_status(self, status: AgentStatus) -> None:
        """Record a status transition and announce it.

        A no-op when the status is unchanged: `agent/status` is a transition
        signal, and emitting it for an assignment would make "settled" mean
        two different things.
        """
        if status is self._status:
            return
        self._status = status
        self._ctx.events.emit(AGENT_STATUS, self, status, scope=self.scope.key)
        if self.on_status_change is not None:
            self.on_status_change(status)

    def reset(self) -> None:
        """Clear runtime state and both queues in place (`AG-016`, `L07-R003`).

        Pinned Pi's `Agent.reset()`, exactly: rejects outright while active
        (`this.activeRun`, matched here by `status is not IDLE`), with the exact
        error text preserved and no partial mutation on rejection. When idle, it
        clears `streamingMessage`/`pendingToolCalls`/`errorMessage` and both
        queues, and retains everything else: object identity, `on_status_change`,
        `definition`, the mutable current `system_prompt`/`model`/`thinking_level`,
        and the tool/scope relationship. This is genuinely in-place -- the same
        `Inbox`/`SessionLog`/scope objects, not a replacement instance -- which is
        the exact property the independent Rust review's rejection turned on.

        One Pi-owned effect is deliberately NOT reproduced: pinned Pi's `reset()`
        also clears `messages` (`this._state.messages = []`). Pi's own transcript
        is a plain in-memory array with no separate persisted log; Minion's
        `SessionLog` (Layer 03, already certified) is append-only by design, with
        no truncate/clear primitive, and adding one -- or teaching `derive_messages`
        to respect a reset boundary -- would be a Layer-03 semantic change this
        narrow Layer-07 remediation has no mandate to make. This is a classified,
        disclosed cross-layer dependency, not a silently dropped requirement: see
        `assurance/layers/07-agent-state-inboxes-python.md`'s PASS 2.5 section.
        """
        if self._status is not AgentStatus.IDLE:
            raise AgentActiveError(
                "Agent is already processing. Wait for completion before resetting."
            )
        self.streaming_message = None
        self.pending_tool_calls = frozenset()
        self.error_message = None
        self.inbox.clear_all()

    def steer(self, message: Message, origin: JsonValue = None) -> InputEnvelope:
        """Pinned Pi's `Agent.steer()` -- the Agent-level public surface
        (`AG-011`), delegating to the authoritative `Inbox`."""
        return self.inbox.steer(message, origin=origin)

    def follow_up(self, message: Message, origin: JsonValue = None) -> InputEnvelope:
        """Pinned Pi's `Agent.followUp()` (`AG-011`)."""
        return self.inbox.followup(message, origin=origin)

    def inject(self, message: Message, origin: JsonValue = None) -> InputEnvelope:
        """Silent context injection -- an intentional Minion extension with no
        Pi equivalent (`AG-011`), exposed at the Agent level alongside `steer`/
        `follow_up` for the same reason those are."""
        return self.inbox.inject(message, origin=origin)

    def has_queued_messages(self) -> bool:
        """Pinned Pi's `Agent.hasQueuedMessages()` -- the Agent-level public
        surface (`AG-011`), delegating to the authoritative `Inbox`."""
        return self.inbox.has_pending()

    def clear_steering_queue(self) -> None:
        """Pinned Pi's `Agent.clearSteeringQueue()` (`AG-013`)."""
        self.inbox.clear(InboxTarget.NEXT_STEP)

    def clear_follow_up_queue(self) -> None:
        """Pinned Pi's `Agent.clearFollowUpQueue()` (`AG-013`)."""
        self.inbox.clear(InboxTarget.NEXT_TURN)

    def clear_all_queues(self) -> None:
        """Pinned Pi's `Agent.clearAllQueues()` (`AG-013`)."""
        self.inbox.clear_all()

File: minion-agent-rust/crates/minion-agent/src/agent/instance.rs
use std::{collections::BTreeSet, sync::Arc};

use parking_lot::Mutex;
use serde_json::Value;
use thiserror::Error;

use crate::{
    agent_loop::{RunConfig, RunContext, RunSnapshot},
    llm::{Message, ModelIdentity},
    runtime::{Context, RunAbortController, RunSignal, ScopeHandle},
    session::{Session, SessionError},
    tools::ToolDefinition,
};

use super::{AgentDefinition, AgentStatus, Inbox, InboxTarget, InputEnvelope, ThinkingLevel};

#[derive(Debug, Eq, Error, PartialEq)]
pub enum AgentError {
    #[error("Agent is already processing. Wait for completion before resetting.")]
    Active,
    #[error("{0}")]
    Session(String),
}

#[derive(Debug, Eq, Error, PartialEq)]
pub enum AgentRunError {
    #[error("Agent is already processing.")]
    Active,
    #[error(transparent)]
    Session(#[from] SessionError),
}

struct RunEntryRollback {
    streaming_message: Option<Message>,
    error_message: Option<String>,
}

#[derive(Clone)]
struct AgentMutableState {
    system_prompt: String,
    model: ModelIdentity,
    thinking_level: ThinkingLevel,
    status: AgentStatus,
    streaming_message: Option<Message>,
    pending_tool_calls: BTreeSet<String>,
    error_message: Option<String>,
    abort_controller: Option<RunAbortController>,
}

pub struct AgentInstance {
    id: String,
    definition: AgentDefinition,
    session: Session,
    context: Option<Context>,
    scope: Option<ScopeHandle>,
    state: Mutex<AgentMutableState>,
    status_gate: Mutex<()>,
    inbox: Inbox,
}

impl AgentInstance {
    pub fn new(
        id: impl Into<String>,
        definition: AgentDefinition,
        session: Session,
        context: Option<Context>,
        scope: Option<ScopeHandle>,
    ) -> Self {
        let state = AgentMutableState {
            system_prompt: definition.system_prompt.clone(),
            model: definition.model.clone(),
            thinking_level: ThinkingLevel::Off,
            status: AgentStatus::Idle,
            streaming_message: None,
            pending_tool_calls: BTreeSet::new(),
            error_message: None,
            abort_controller: None,
        };
        Self {
            id: id.into(),
            definition,
            session,
            context,
            scope,
            state: Mutex::new(state),
            status_gate: Mutex::new(()),
            inbox: Inbox::new(),
        }
    }

    pub fn id(&self) -> &str {
        &self.id
    }
    pub fn definition(&self) -> &AgentDefinition {
        &self.definition
    }
    pub fn session(&self) -> &Session {
        &self.session
    }
    pub fn inbox(&self) -> &Inbox {
        &self.inbox
    }

    pub fn steer(&self, message: Message, origin: Option<Value>) -> InputEnvelope {
        self.inbox.steer(message, origin)
    }

    pub fn follow_up(&self, message: Message, origin: Option<Value>) -> InputEnvelope {
        self.inbox.follow_up(message, origin)
    }

    pub fn inject(&self, message: Message, origin: Option<Value>) -> InputEnvelope {
        self.inbox.inject(message, origin)
    }

    pub fn clear_steering_queue(&self) {
        self.inbox.clear(InboxTarget::Steering);
    }

    pub fn clear_follow_up_queue(&self) {
        self.inbox.clear(InboxTarget::FollowUp);
    }

    pub fn clear_all_queues(&self) {
        self.inbox.clear_all();
    }

    pub fn has_queued_messages(&self) -> bool {
        self.inbox.has_pending()
    }
    pub fn system_prompt(&self) -> String {
        self.state.lock().system_prompt.clone()
    }
    pub fn model(&self) -> ModelIdentity {
        self.state.lock().model.clone()
    }
    pub fn thinking_level(&self) -> ThinkingLevel {
        self.state.lock().thinking_level
    }
    pub fn status(&self) -> AgentStatus {
        self.state.lock().status
    }
    pub fn streaming_message(&self) -> Option<Message> {
        self.state.lock().streaming_message.clone()
    }
    pub fn pending_tool_calls(&self) -> BTreeSet<String> {
        self.state.lock().pending_tool_calls.clone()
    }
    pub fn error_message(&self) -> Option<String> {
        self.state.lock().error_message.clone()
    }

    /// Returns the read-only cancellation signal for the active run.
    pub fn signal(&self) -> Option<RunSignal> {
        self.state
            .lock()
            .abort_controller
            .as_ref()
            .map(RunAbortController::signal)
    }

    /// Requests cooperative cancellation. Calling this while idle is a no-op.
    pub fn abort(&self) {
        if let Some(controller) = self.state.lock().abort_controller.as_ref() {
            controller.abort();
        }
    }

    pub fn set_system_prompt(&self, value: impl Into<String>) {
        self.state.lock().system_prompt = value.into();
    }
    pub fn set_model(&self, value: ModelIdentity) {
        self.state.lock().model = value;
    }
    pub fn set_thinking_level(&self, value: ThinkingLevel) {
        self.state.lock().thinking_level = value;
    }
    pub fn set_streaming_message(&self, value: Option<Message>) {
        self.state.lock().streaming_message = value;
    }
    pub fn set_pending_tool_calls(&self, value: BTreeSet<String>) {
        self.state.lock().pending_tool_calls = value;
    }
    pub fn set_error_message(&self, value: Option<String>) {
        self.state.lock().error_message = value;
    }

    pub fn set_status(&self, status: AgentStatus) {
        let _gate = self.status_gate.lock();
        self.state.lock().status = status;
    }

    pub fn try_begin_run(&self) -> Result<RunSnapshot, AgentRunError> {
        let (system_prompt, model, thinking_level, signal, rollback) = {
            let _gate = self.status_gate.lock();
            let mut state = self.state.lock();
            if state.status != AgentStatus::Idle {
                return Err(AgentRunError::Active);
            }
            let rollback = RunEntryRollback {
                streaming_message: state.streaming_message.clone(),
                error_message: state.error_message.clone(),
            };
            let controller = RunAbortController::default();
            let signal = controller.signal();
            state.abort_controller = Some(controller);
            state.status = AgentStatus::Running;
            state.streaming_message = None;
            state.error_message = None;
            (
                state.system_prompt.clone(),
                state.model.clone(),
                state.thinking_level,
                signal,
                rollback,
            )
        };

        let messages = match self.session.derive_messages() {
            Ok(messages) => messages,
            Err(error) => {
                self.rollback_run_entry(rollback);
                return Err(error.into());
            }
        };
        let tools = self.tools();

        Ok(RunSnapshot {
            context: RunContext {
                system_prompt,
                messages,
                tools,
            },
            config: RunConfig {
                model,
                thinking_level,
            },
            signal,
        })
    }

    pub fn finish_run(&self) {
        let _gate = self.status_gate.lock();
        let mut state = self.state.lock();
        state.abort_controller = None;
        state.status = AgentStatus::Idle;
        state.streaming_message = None;
        state.pending_tool_calls.clear();
    }

    pub(crate) fn add_pending_tool_call(&self, tool_call_id: String) {
        self.state.lock().pending_tool_calls.insert(tool_call_id);
    }

    pub(crate) fn remove_pending_tool_call(&self, tool_call_id: &str) {
        self.state.lock().pending_tool_calls.remove(tool_call_id);
    }

    fn rollback_run_entry(&self, rollback: RunEntryRollback) {
        let _gate = self.status_gate.lock();
        let mut state = self.state.lock();
        state.abort_controller = None;
        state.status = AgentStatus::Idle;
        state.streaming_message = rollback.streaming_message;
        state.error_message = rollback.error_message;
    }

    pub fn messages(&self) -> Result<Vec<Message>, AgentError> {
        self.session
            .derive_messages()
            .map_err(|error| AgentError::Session(error.to_string()))
    }

    pub fn tools(&self) -> Vec<Arc<ToolDefinition>> {
        self.context
            .as_ref()
            .and_then(|context| context.tools().ok())
            .map_or_else(Vec::new, |registry| registry.visible(self.scope.as_ref()))
    }

    pub fn reset(&self) -> Result<(), AgentError> {
        let _gate = self.status_gate.lock();
        if self.state.lock().status != AgentStatus::Idle {
            return Err(AgentError::Active);
        }
        self.session
            .reset()
            .map_err(|error| AgentError::Session(error.to_string()))?;
        {
            let mut state = self.state.lock();
            state.streaming_message = None;
            state.pending_tool_calls.clear();
            state.error_message = None;
        }
        self.inbox.clear_all();
        Ok(())
    }
}
